modules/frontend/pipeline/async_sharding.go

```diff
@@ -0,0 +1,104 @@
+package pipeline
+
+import (
+	"net/http"
+	"sync"
+
+	"github.com/grafana/tempo/pkg/boundedwaitgroup"
+)
+
+type waitGroup interface {
+	Add(int)
+	Done()
+	Wait()
+}
+
+// NewAsyncSharder creates a new AsyncResponse that shards requests to the next AsyncRoundTripper[*http.Response]. It creates one
+// goroutine per concurrent request.
+func NewAsyncSharder(concurrent int, reqFn func(i int) *http.Request, next AsyncRoundTripper[*http.Response]) Responses[*http.Response] {
+	var wg waitGroup
+	if concurrent <= 0 {
+		wg = &sync.WaitGroup{}
+	} else {
+		bwg := boundedwaitgroup.New(uint(concurrent))
+		wg = &bwg
+	}
+	asyncResp := newAsyncResponse()
+
+	go func() {
+		defer asyncResp.done()
+
+		reqIdx := 0
+		for {
+			req := reqFn(reqIdx)
+			reqIdx++
+
+			// else check for a request to pass down the pipeline
+			if req == nil {
+				break
+			}
+
+			wg.Add(1)
+			go func(r *http.Request) {
+				defer wg.Done()
+
+				resp, err := next.RoundTrip(r)
+				if err != nil {
+					asyncResp.SendError(err)
+					return
+				}
+
+				asyncResp.Send(resp)
+			}(req)
+		}
+
+		wg.Wait()
+	}()
+
+	return asyncResp
+}
+
+// NewAsyncSharderLimitedGoroutines creates a new AsyncResponse that shards requests to the next AsyncRoundTripper[*http.Response] using a limited number of goroutines.
+func NewAsyncSharderLimitedGoroutines(concurrent int, reqs <-chan *http.Request, resps Responses[*http.Response], next AsyncRoundTripper[*http.Response]) Responses[*http.Response] {
```

Comment:
These two methods would be clearer to me if they were renamed to `NewAsyncSharderFunc` and `NewAsyncSharderChan`.  That looks like the main difference (in this current version at least). Since they both have concurrently limits.

Reply:
The intention was for `NewAsyncSharder` to create one goroutine for every concurrrent downstream request. This is the way the current frontend works.

`NewAsyncSharderLimitedGoroutines` creates a limited number of goroutine "workers" that can then fire up to concurrency requests downstream.

I don't think this is working the way I intended it to. This was the last area of code I was looking at closely. I'm going to do some more experimentation but I think that we will still need one goroutine per request. Perhaps we eventually migrate to an async request pipeline as well as response?

Reply:
Ok, I've cleaned these up and used your suggested names. I like where they are now, but think there's room for improvement.